### PR TITLE
refactor(anchor): use new Anchor error features, add more JS integration tests, add client-side EventQueue schema

### LINKED ISSUE #15: better errors
Anchor's latest release (0.22) introduces some very handy new features and ergonomics for error handling. we can leverage them in the AOB now that it's using Anchor

### PR DESCRIPTION
**Pull request by lidatong**
Closes #15

# Summary
After discussions with Jarry and Armani, we reached consensus we're going to try out a "static sized" buffer. Essentially, having dynamically sized types (e.g. `event_queue`) is a huge headache:
- limitations with `zero-copy` - `bytemuck` is restricted to `Sized` types (other than the special case of `[T]`).
- inability to leverage `Anchor` codegen and IDL.

- the cost of course, is the user loses flexibility to decide exactly how much they want to size their AOB types (and will likely overallocate).
- having a choice of sizes (e.g. `S / M / L` in terms of # of bytes needed) could help alleviate this... but I think none of us agree it's the "best" solution.

I think the "right" solution is `bytemuck` supports user DSTs and Anchor can codegen with those same DSTs, which I think is planned (this is more punting _for now_).

# Other notes
As part of this PR, I also needed to relocate the AOB code for the sake of the IDL generation... need to discuss whether it's possible to keep them separate and still generate IDL types.

### CODE REVIEW

File: anchor/programs/anchor-agnostic-orderbook/src/lib.rs [hunk 1]
```diff
@@ -309,8 +283,8 @@ pub struct CreateMarket<'info> {
     #[account(init, payer = payer)]
     pub market: AccountLoader<'info, MarketState>,
     // TODO pass in space size instead of just max
-    #[account(init, payer = payer, space = 10240)]
-    pub event_queue: AccountInfo<'info>,
+    #[account(init, payer = payer, space = 8 + 591)]
```
[the remark this answers is not in the document]
Comment (lidatong) [reply anchor/programs/anchor-agnostic-orderbook/src/lib.rs:286]: yes, good suggestion

File: anchor/programs/anchor-agnostic-orderbook/src/aob/error.rs [hunk 2]
```diff
@@ -1,7 +1,7 @@
 use num_derive::FromPrimitive;
 use num_traits::FromPrimitive;
-use solana_program::program_error::PrintProgramError;
-use solana_program::{decode_error::DecodeError, msg, program_error::ProgramError};
+use anchor_lang::solana_program::program_error::PrintProgramError;
+use anchor_lang::solana_program::{decode_error::DecodeError, msg, program_error::ProgramError};
```
[the remark this answers is not in the document]
Comment (lidatong) [reply anchor/programs/anchor-agnostic-orderbook/src/aob/error.rs:4]: `anchor_lang` re-exports `solana_program` so no need to depend on it directly

File: anchor/programs/anchor-agnostic-orderbook/src/aob/state.rs [hunk 3]
```diff
@@ -287,55 +283,18 @@ impl EventQueueHeader {
...
 ///
 /// This struct is used at runtime but doesn't represent a serialized event queue
-pub struct EventQueue<'a> {
+#[account]
+#[derive(Debug)]
+pub struct EventQueue {
     pub header: EventQueueHeader,
-    pub(crate) buffer: Rc<RefCell<&'a mut [u8]>>, //The whole account data
-    callback_info_len: usize,
+    pub buffer: [u8; 512], //The whole account data
```
Comment (jarry-xiao) [inline anchor/programs/anchor-agnostic-orderbook/src/aob/state.rs:290]: this is probably too small

**Comment by lidatong** [comment]
looked into `heapless` as a potential solution, but unfortunately it does not implement `Copy` (i believe this is because their data structures have to impl `Drop`, e.g. `heapless::Vec<ptrToHeapData>` is also valid)

this precludes us from zero-copy because bytemuck has a `Pod: Copy` constraint. the reason is `Copy` is a bytemuck's proxy for `not Drop` (https://github.com/Lokathor/bytemuck/issues/87)

the problem is Rust does not give you any way to express `unsafe impl Copy` -- it's a frustrating limitation, because i can _know_ a type is `Copy` (does not drop) yet have no way to tell the type checker that

it would have been nice to use a standard data structure library like `heapless` along with Anchor IDL and bytemuck's zero-copy but it doesn't seem that's an option (at least right now)

**Comment by lidatong** [comment]
update:
- instead of `heapless`, i'm going with `[EventQueue; N]` (it'd be nice to use a generic collections crate, but the circular buffer API is simple enough)
- been wrestling with Anchor and `zero_copy`. after a fair bit of debugging, i realized these seem to be bugs with the IDL generation with respect to zero-copy alignment and how it composes with `Option` / `Enum`
- as an example, I was seeing random errors like this:

```
  1) anchor-agnostic-orderbook
       consume events:
     Error: Invalid bool: 211
```

whereas a Rust client `bytemuck::from_bytes` would work just fine

searching github, it seems others are running into these issues too:
- https://github.com/project-serum/anchor/issues/806
- https://github.com/project-serum/anchor/issues/904
- i have idea on how to restructure around these limitations (not ideal, but can do for now)

**_to be clear, i still like Anchor_** -- `zero_copy` is just less popular than Borsh, making for few examples and more gnarly to debug. but now that i've run into these i can save everyone else time.

**Comment by lidatong** [comment]
unfortunately `[T; N]` (const generic) was also a bust, it panics with `thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: ParseIntError { kind: InvalidDigit }', lang/syn/src/idl/mod.rs:202:76` (Anchor can't parse it)

**Review by leina05 — APPROVED** [review]
This all looks good. Would be good to update the PR name/comment to be more up to date on what's actually going in with these changes, which seems to be some refactoring to use new Anchor errors and some client-side types for deserializing `EventQueue` accounts.

To summarize my current understanding, the next things we want to tackle include: 
- `Slab` and `EventQueue` both use Borsh; it would be good to make these zero copy somehow
- We need to benchmark compute unit consumption against dex-v3, and think about what our targets are...a good concrete goal is probably to implement a "cancel all orders" instruction that actually works, i.e. https://github.com/Bonfida/agnostic-orderbook/issues/17
- Improve error handling using new Anchor error features

File: anchor/tests/eventQueue.ts [hunk 4]
```diff
@@ -0,0 +1,284 @@
+import {Connection, PublicKey} from "@solana/web3.js";
```
[the remark this answers is not in the document]
Comment (lidatong) [reply anchor/tests/eventQueue.ts:1]: the import or the file?

File: anchor/programs/anchor-agnostic-orderbook/src/aob/histbuf.rs [hunk 5]
```diff
@@ -0,0 +1,40 @@
+/// A simplified port of heapless' "history buffer" with `Pod` and `Default` constraints
```
[the remark this answers is not in the document]
Comment (lidatong) [reply anchor/programs/anchor-agnostic-orderbook/src/aob/histbuf.rs:1]: removed, thanks

File: program/src/processor/close_market.rs [hunk 6]
```diff
@@ -3,7 +3,7 @@ use aob::params::CloseMarketParams;
 use aob::{
```
[the remark this answers is not in the document]
Comment (lidatong) [reply program/src/processor/close_market.rs:3]: yeah i thought about the same thing. main thing w/ separate repo is it becomes harder to merge back upstream

**Comment by lidatong** [comment]
> Would be good to update the PR name/comment

thanks, done

> Slab and EventQueue both use Borsh

in both cases just the headers. the actual buffer is zero-copy via `Rc::clone`

> benchmark compute unit

agree on benchmarking, but i think a fair comparison will be:
- current anchor AOB vs. upstream AOB
- anchor dex-v4 vs. upstream dex-v4 vs. serum-dex
- i expect serum-dex to win because they won't have the CPI overhead of using an AOB for matching. but will be good to have numbers

**Comment by lidatong** [comment]
i think some good next steps before merge:
- [x] properly refactoring to use Anchor's new errors

edit: moved to issues

**Pull request merged**